In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
dir1 = r"/Users/zhilinyang/Desktop/data_Q1_2017"
dir2 = r"/Users/zhilinyang/Desktop/data_Q2_2017"
dir3 = r"/Users/zhilinyang/Desktop/data_Q3_2017"
dir4 = r"/Users/zhilinyang/Desktop/data_Q4_2017"
root_dir=[dir1,dir2,dir3,dir4]

In [3]:
basic_features=['smart_1_raw','smart_3_normalized','smart_5_normalized','smart_7_normalized',
         'smart_9_normalized','smart_187_normalized','smart_189_normalized',
          'smart_194_normalized','smart_197_normalized',
          'smart_5_raw','smart_197_raw']
colname=['serial_number','failure']+basic_features

In [9]:
for i in range(0,3): 
    for file in os.listdir(root_dir[i]):
        file_name = root_dir[i] + '/' + file
        df = pd.read_csv(file_name)

        #Select all Seagate A samples and its features
        dfST = df[df['model'].str.match('ST4000DM000')]
        dfST = dfST[colname]
        
        #combine all dates' data into a single file
        dfST.to_csv(path_or_buf=r"/Users/zhilinyang/Desktop/STA.csv", mode='a')
        print(file)

2017-01-01.csv
2017-01-02.csv
2017-01-03.csv
2017-01-04.csv
2017-01-05.csv
2017-01-06.csv
2017-01-07.csv
2017-01-08.csv
2017-01-09.csv
2017-01-10.csv
2017-01-11.csv
2017-01-12.csv
2017-01-13.csv
2017-01-14.csv
2017-01-15.csv
2017-01-16.csv
2017-01-17.csv
2017-01-18.csv
2017-01-19.csv
2017-01-20.csv
2017-01-21.csv
2017-01-22.csv
2017-01-23.csv
2017-01-24.csv
2017-01-25.csv
2017-01-26.csv
2017-01-27.csv
2017-01-28.csv
2017-01-29.csv
2017-01-30.csv
2017-01-31.csv
2017-02-01.csv
2017-02-02.csv
2017-02-03.csv
2017-02-04.csv
2017-02-05.csv
2017-02-06.csv
2017-02-07.csv
2017-02-08.csv
2017-02-09.csv
2017-02-10.csv
2017-02-11.csv
2017-02-12.csv
2017-02-13.csv
2017-02-14.csv
2017-02-15.csv
2017-02-16.csv
2017-02-17.csv
2017-02-18.csv
2017-02-19.csv
2017-02-20.csv
2017-02-21.csv
2017-02-22.csv
2017-02-23.csv
2017-02-24.csv
2017-02-25.csv
2017-02-26.csv
2017-02-27.csv
2017-02-28.csv
2017-03-01.csv
2017-03-02.csv
2017-03-03.csv
2017-03-04.csv
2017-03-05.csv
2017-03-06.csv
2017-03-07.csv
2017-03-08

In [10]:
omit=True
for file in os.listdir(root_dir[3]):
        if omit:
            omit=False
            continue
        file_name = root_dir[3] + '/' + file
        df = pd.read_csv(file_name)

        #Select all Seagate A samples and its features
        dfST = df[df['model'].str.match('ST4000DM000')]
        dfST = dfST[colname]
        
        #combine all dates' data into a single file
        dfST.to_csv(path_or_buf=r"/Users/zhilinyang/Desktop/STA.csv", mode='a')
        print(file)

2017-10-01.csv
2017-10-02.csv
2017-10-03.csv
2017-10-04.csv
2017-10-05.csv
2017-10-06.csv
2017-10-07.csv
2017-10-08.csv
2017-10-09.csv
2017-10-10.csv
2017-10-11.csv
2017-10-12.csv
2017-10-13.csv
2017-10-14.csv
2017-10-15.csv
2017-10-16.csv
2017-10-17.csv
2017-10-18.csv
2017-10-19.csv
2017-10-20.csv
2017-10-21.csv
2017-10-22.csv
2017-10-23.csv
2017-10-24.csv
2017-10-25.csv
2017-10-26.csv
2017-10-27.csv
2017-10-28.csv
2017-10-29.csv
2017-10-30.csv
2017-10-31.csv
2017-11-01.csv
2017-11-02.csv
2017-11-03.csv
2017-11-04.csv
2017-11-05.csv
2017-11-06.csv
2017-11-07.csv
2017-11-08.csv
2017-11-09.csv
2017-11-10.csv
2017-11-11.csv
2017-11-12.csv
2017-11-13.csv
2017-11-14.csv
2017-11-15.csv
2017-11-16.csv
2017-11-17.csv
2017-11-18.csv
2017-11-19.csv
2017-11-20.csv
2017-11-21.csv
2017-11-22.csv
2017-11-23.csv
2017-11-24.csv
2017-11-25.csv
2017-11-26.csv
2017-11-27.csv
2017-11-28.csv
2017-11-29.csv
2017-11-30.csv
2017-12-01.csv
2017-12-02.csv
2017-12-03.csv
2017-12-04.csv
2017-12-05.csv
2017-12-06

In [5]:
stdf=stdf[stdf['failure']!='failure']
stdf=stdf.drop('smart_195_normalized',axis=1)
stdf=stdf.dropna()
stdf.head()

serial_number failure smart_1_raw smart_3_normalized smart_5_normalized  \
0      Z305B2QN       0    58173272                 91                100   
1      Z302A0YH       0    75626904                 92                100   
2      Z305BT0W       0    48893128                 93                100   
3      Z302A0YE       0   192617016                 92                100   
4      Z302PGH8       0    83211536                 92                100   

  smart_7_normalized smart_9_normalized smart_187_normalized  \
0                 85                 90                100.0   
1                 88                 81                100.0   
2                 84                 92                100.0   
3                 75                 81                100.0   
4                 87                 85                100.0   

  smart_189_normalized smart_194_normalized smart_197_normalized smart_5_raw  \
0                100.0                   24                  100           0   
1                100.0                   19                  100           0   
2                100.0                   30                  100           0   
3                100.0                   21                  100           0   
4                100.0                   25                  100           0   

  smart_197_raw  
0             0  
1             0  
2             0  
3             0  
4             0

In [16]:
stdf.isnull().sum()

serial_number           0
failure                 0
smart_1_raw             0
smart_3_normalized      0
smart_5_normalized      0
smart_7_normalized      0
smart_9_normalized      0
smart_187_normalized    0
smart_189_normalized    0
smart_194_normalized    0
smart_197_normalized    0
smart_5_raw             0
smart_197_raw           0
dtype: int64

In [6]:
print('total num of disks: %d' % len(stdf['serial_number'].unique()))

total num of disks: 35189


In [7]:
print('failed disks: %d' % (pd.to_numeric(stdf['failure'])).sum())

failed disks: 1061


In [11]:
failed_disks=stdf[stdf['failure']=='1']
failed_disks

serial_number failure  smart_1_raw smart_3_normalized  \
31020         Z3015LED       1    162296184                 91   
36991         Z3004D6W       1    243309976                 92   
46843         Z300GPL4       1     37445368                 92   
53417         Z300CNVT       1    151213056                 92   
57396         S300X9C1       1    147629472                 93   
57705         Z300NKVG       1     44940944                 91   
88617         S300XR7S       1    151413392                 92   
89011         Z3041CZ5       1    156786872                 97   
118457        Z300X5E2       1    159236192                 94   
150196        S300Z4JV       1     52699768                 98   
150972        Z302FR4F       1     22908872                 97   
183320        Z3016SRV       1     65231360                 91   
185099        Z3023XGJ       1     76721336                 97   
197060        Z3029AZC       1     12014880                 96   
207023        Z304L8TB       1    235405080                 98   
207696        Z300GQAS       1     62117488                 92   
210292        Z300KCQB       1    236668304                 91   
309580        W300E0B9       1     18608976                 92   
350364        Z300ZV76       1      9493888                 94   
364395        Z300MJP5       1    128439688                 91   
375018        Z302A0ZV       1     22015008                 95   
408358        S300YTW4       1    209638920                 92   
418720        Z300KCQV       1    231616384                 92   
426855        Z3015BZD       1    182861968                 92   
489184        Z300WXNR       1     55642624                 91   
496047        Z302B0A5       1     73876912                 91   
496737        Z300Y4NH       1    209296064                 91   
522462        W300EK8X       1    143882040                 93   
548101        S3010CM5       1    187610424                 92   
553373        Z302AKVP       1     42257736                 92   
...                ...     ...          ...                ...   
11814090      Z3041857       1     34623672                 91   
11819677      Z302515G       1    144791160                 92   
11845870      Z305D5ME       1     64013336                 93   
11849410      S300VL64       1    121051161                 93   
11865889      S301F95D       1    233843578                 96   
11889987      Z304KBFE       1    179188832                 92   
11893452      Z3011WAL       1    190209488                 91   
11895919      Z300ZV58       1     71690288                 91   
11931077      Z300YV3N       1     22338480                 91   
11931161      Z302SYPC       1     49497266                 92   
11980827      Z300GQ6V       1      1036576                 92   
11980975      Z30149KC       1    161613688                 92   
11982964      Z3029A8B       1  224810088.0               92.0   
11987637      W300AJHL       1   58894088.0               91.0   
12006302      S30101R4       1  111908400.0               93.0   
12007038      Z305AEVK       1  184245368.0               93.0   
12008693      Z304KBNL       1   48758904.0               92.0   
12011277      Z304GHS5       1  174750800.0               96.0   
12018672      Z3014P0L       1    155354624                 91   
12048063      Z3029GV9       1    215068558                 91   
12048216      S301GNPF       1    132017152                 96   
12064943      W300H5H6       1    236006176                 91   
12077555      Z30250HT       1    118178960                 93   
12095882      S300Z0TD       1     95480384                 94   
12100220      Z300V5RQ       1     19799904                 92   
12114558      Z305BAG3       1    207913144                 94   
12116146      Z3023TWW       1    239873512                 91   
12152693      Z305BHVV       1    186154848                 92   
12211110      Z300GZ1B       1     17846936      

In [26]:
len(failed_disks['serial_number'].unique())

1058

There are totally 1058 unique failed disks, where several of them failed twice
So totally 1061 failed disks

In [18]:
fail_name_list=failed_disks['serial_number']
stdf.index=stdf['serial_number']
stdf.head()

serial_number failure smart_1_raw smart_3_normalized  \
serial_number                                                        
Z305B2QN           Z305B2QN       0    58173272                 91   
Z302A0YH           Z302A0YH       0    75626904                 92   
Z305BT0W           Z305BT0W       0    48893128                 93   
Z302A0YE           Z302A0YE       0   192617016                 92   
Z302PGH8           Z302PGH8       0    83211536                 92   

              smart_5_normalized smart_7_normalized smart_9_normalized  \
serial_number                                                            
Z305B2QN                     100                 85                 90   
Z302A0YH                     100                 88                 81   
Z305BT0W                     100                 84                 92   
Z302A0YE                     100                 75                 81   
Z302PGH8                     100                 87                 85   

              smart_187_normalized smart_189_normalized smart_194_normalized  \
serial_number                                                                  
Z305B2QN                     100.0                100.0                   24   
Z302A0YH                     100.0                100.0                   19   
Z305BT0W                     100.0                100.0                   30   
Z302A0YE                     100.0                100.0                   21   
Z302PGH8                     100.0                100.0                   25   

              smart_197_normalized smart_5_raw smart_197_raw  
serial_number                                                 
Z305B2QN                       100           0             0  
Z302A0YH                       100           0             0  
Z305BT0W                       100           0             0  
Z302A0YE                       100           0             0  
Z302PGH8                       100           0             0

In [21]:
stdf['serial_number']=stdf['serial_number'].isin(fail_name_list)
stdf.head()

serial_number failure smart_1_raw smart_3_normalized  \
serial_number                                                         
Z305B2QN               False       0    58173272                 91   
Z302A0YH               False       0    75626904                 92   
Z305BT0W               False       0    48893128                 93   
Z302A0YE               False       0   192617016                 92   
Z302PGH8               False       0    83211536                 92   

              smart_5_normalized smart_7_normalized smart_9_normalized  \
serial_number                                                            
Z305B2QN                     100                 85                 90   
Z302A0YH                     100                 88                 81   
Z305BT0W                     100                 84                 92   
Z302A0YE                     100                 75                 81   
Z302PGH8                     100                 87                 85   

              smart_187_normalized smart_189_normalized smart_194_normalized  \
serial_number                                                                  
Z305B2QN                     100.0                100.0                   24   
Z302A0YH                     100.0                100.0                   19   
Z305BT0W                     100.0                100.0                   30   
Z302A0YE                     100.0                100.0                   21   
Z302PGH8                     100.0                100.0                   25   

              smart_197_normalized smart_5_raw smart_197_raw  
serial_number                                                 
Z305B2QN                       100           0             0  
Z302A0YH                       100           0             0  
Z305BT0W                       100           0             0  
Z302A0YE                       100           0             0  
Z302PGH8                       100           0             0

In [22]:
failed_samples=stdf[stdf['serial_number']==True]

In [33]:
failed_samples.loc['Z300KMM3']

serial_number failure smart_1_raw smart_3_normalized  \
serial_number                                                         
Z300KMM3                True       0   221434504                 91   
Z300KMM3                True       0   143146672                 91   
Z300KMM3                True       0    23451136                 91   
Z300KMM3                True       0   142618120                 91   
Z300KMM3                True       0    84155840                 91   
Z300KMM3                True       0    16426224                 91   
Z300KMM3                True       0   130843336                 91   
Z300KMM3                True       0   207033704                 91   
Z300KMM3                True       0   131373928                 91   
Z300KMM3                True       0    37864536                 91   
Z300KMM3                True       0   182203712                 91   
Z300KMM3                True       0    99255680                 91   
Z300KMM3                True       0    24155344                 91   
Z300KMM3                True       0   163390320                 91   
Z300KMM3                True       0    77501488                 91   
Z300KMM3                True       0    35880704                 91   
Z300KMM3                True       0   203904008                 91   
Z300KMM3                True       0    88538480                 91   
Z300KMM3                True       0    39500336                 91   
Z300KMM3                True       0   175589824                 91   
Z300KMM3                True       0   105608208                 91   
Z300KMM3                True       0   240825376                 91   
Z300KMM3                True       0    91751536                 91   
Z300KMM3                True       0    10588040                 91   
Z300KMM3                True       0   149492856                 91   
Z300KMM3                True       0    58942936                 91   
Z300KMM3                True       0    58942936                 91   
Z300KMM3                True       0   167307320                 91   
Z300KMM3                True       0   195762688                 91   
Z300KMM3                True       0     8768928                 91   
...                      ...     ...         ...                ...   
Z300KMM3                True       0   111664312                 91   
Z300KMM3                True       0    37481424                 91   
Z300KMM3                True       0   174135272                 91   
Z300KMM3                True       0    48858912                 91   
Z300KMM3                True       0   165497648                 91   
Z300KMM3                True       0    95478104                 91   
Z300KMM3                True       0   209041056                 91   
Z300KMM3                True       0   149178344                 91   
Z300KMM3                True       0    79697992                 91   
Z300KMM3                True       0    21083264                 91   
Z300KMM3                True       0   220580704                 91   
Z300KMM3                True       0   115255232                 91   
Z300KMM3                True       0     7284112                 91   
Z300KMM3                True       0   186681088                 91   
Z300KMM3                True       0    73892840                 91   
Z300KMM3                True       0   198702600                 91   
Z300KMM3                True       0    79371312                 91   
Z300KMM3                True       0     7768648                 91   
Z300KMM3                True       0   129752480                 91   
Z300KMM3                True       0     4471384                 91   
Z300KMM3                True       0   155593208                 91   
Z300KMM3                True       0    37454672                 91   
Z300KMM3                True       0   118520200                 91   
Z300KMM3                True       0   

In [34]:
failed_samples.to_csv('/Users/zhilinyang/Desktop/failed_samples.csv')
print('done')

done


In [35]:
stdf=stdf.drop('serial_number',axis=1)
stdf.to_csv('/Users/zhilinyang/Desktop/STA.csv')
print('done')

done


In [2]:
stdf=pd.read_csv('/Users/zhilinyang/Desktop/work/STA.csv')
stdf.head()

serial_number  failure  smart_1_raw  smart_3_normalized  smart_5_normalized  \
0      Z305B2QN        0   58173272.0                91.0               100.0   
1      Z302A0YH        0   75626904.0                92.0               100.0   
2      Z305BT0W        0   48893128.0                93.0               100.0   
3      Z302A0YE        0  192617016.0                92.0               100.0   
4      Z302PGH8        0   83211536.0                92.0               100.0   

   smart_7_normalized  smart_9_normalized  smart_187_normalized  \
0                85.0                90.0                 100.0   
1                88.0                81.0                 100.0   
2                84.0                92.0                 100.0   
3                75.0                81.0                 100.0   
4                87.0                85.0                 100.0   

   smart_189_normalized  smart_194_normalized  smart_197_normalized  \
0                 100.0                  24.0                 100.0   
1                 100.0                  19.0                 100.0   
2                 100.0                  30.0                 100.0   
3                 100.0                  21.0                 100.0   
4                 100.0                  25.0                 100.0   

   smart_5_raw  smart_197_raw  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0

In [17]:
final_features=['smart_1_raw','smart_3_normalized','smart_5_normalized','smart_7_normalized',
         'smart_9_normalized','smart_187_normalized','smart_189_normalized',
          'smart_194_normalized','smart_195_normalized',
          'smart_5_raw']

In [16]:
fail_df=pd.read_csv('/Users/zhilinyang/Desktop/work/failed_samples.csv')
fail_df.head()

serial_number  serial_number.1  failure  smart_1_raw  smart_3_normalized  \
0      Z300KMM3             True        0  221434504.0                91.0   
1      Z302DK2R             True        0  109378928.0                92.0   
2      Z304H3GQ             True        0   60343088.0                95.0   
3      Z305AT6T             True        0  229967584.0                99.0   
4      Z3051NNA             True        0  215001112.0                92.0   

   smart_5_normalized  smart_7_normalized  smart_9_normalized  \
0               100.0                81.0                67.0   
1               100.0                87.0                83.0   
2               100.0                84.0                91.0   
3               100.0                85.0                90.0   
4               100.0                85.0                90.0   

   smart_187_normalized  smart_189_normalized  smart_194_normalized  \
0                 100.0                  96.0                  27.0   
1                  79.0                 100.0                  24.0   
2                 100.0                 100.0                  26.0   
3                 100.0                 100.0                  22.0   
4                 100.0                 100.0                  29.0   

   smart_197_normalized  smart_5_raw  smart_197_raw  
0                 100.0          0.0            0.0  
1                 100.0          8.0            8.0  
2                 100.0          0.0            0.0  
3                 100.0          0.0            0.0  
4                 100.0          0.0            0.0

In [12]:
len(stdf)

12237697

In [13]:
len(fail_df)

182070

For failed samples: Use all drives and divide them randomly into 7:3

In [14]:
fail_disk_list=fail_df[fail_df['failure']==1]
len(fail_disk_list)

1061

In [11]:
train_disk_list=fail_disk_list.sample(frac=0.7)
test_disk_list=fail_disk_list[~fail_disk_list['serial_number'].isin(train_disk_list['serial_number'])]

In [12]:
train_disk_list.head()

serial_number  serial_number.1  failure  smart_1_raw  \
177531      Z3016WC8             True        1   56820384.0   
21118       Z300J9C4             True        1  179359256.0   
156285      Z304L8TV             True        1  134649210.0   
73367       Z305GVTM             True        1  179660944.0   
122751      Z304KBXJ             True        1  144818456.0   

        smart_3_normalized  smart_5_normalized  smart_7_normalized  \
177531                93.0               100.0                89.0   
21118                 92.0               100.0                89.0   
156285                93.0               100.0                88.0   
73367                 98.0               100.0                86.0   
122751                91.0               100.0                87.0   

        smart_9_normalized  smart_187_normalized  smart_189_normalized  \
177531                64.0                 100.0                  96.0   
21118                 67.0                 100.0                  99.0   
156285                83.0                  81.0                 100.0   
73367                 90.0                 100.0                 100.0   
122751                85.0                  93.0                 100.0   

        smart_194_normalized  smart_197_normalized  smart_5_raw  smart_197_raw  
177531                  20.0                  99.0          0.0          216.0  
21118                   22.0                 100.0          0.0            0.0  
156285                  21.0                 100.0          0.0            8.0  
73367                   26.0                 100.0         16.0            8.0  
122751                  22.0                 100.0          0.0            8.0

In [13]:
test_disk_list.head()

serial_number  serial_number.1  failure  smart_1_raw  smart_3_normalized  \
1615      Z300CNVT             True        1  151213056.0                92.0   
1731      S300X9C1             True        1  147629472.0                93.0   
1738      Z300NKVG             True        1   44940944.0                91.0   
2680      Z3041CZ5             True        1  156786872.0                97.0   
4530      Z302FR4F             True        1   22908872.0                97.0   

      smart_5_normalized  smart_7_normalized  smart_9_normalized  \
1615               100.0                89.0                67.0   
1731               100.0                88.0                81.0   
1738               100.0                83.0                70.0   
2680               100.0                87.0                88.0   
4530               100.0                87.0                83.0   

      smart_187_normalized  smart_189_normalized  smart_194_normalized  \
1615                  98.0                  99.0                  26.0   
1731                  99.0                 100.0                  28.0   
1738                 100.0                  96.0                  22.0   
2680                 100.0                 100.0                  27.0   
4530                  82.0                 100.0                  24.0   

      smart_197_normalized  smart_5_raw  smart_197_raw  
1615                 100.0          8.0            8.0  
1731                 100.0          0.0            0.0  
1738                 100.0          0.0           56.0  
2680                 100.0          0.0            0.0  
4530                 100.0          0.0            8.0

In [14]:
len(train_disk_list)+len(test_disk_list)

1060

In [40]:
grouped=stdf.groupby('serial_number')
grouped.get_group('Z3016SRV')

serial_number  failure  smart_1_raw  smart_3_normalized  \
9633        Z3016SRV        0   55600176.0                91.0   
44370       Z3016SRV        0   26790736.0                91.0   
79105       Z3016SRV        0    4721200.0                91.0   
113841      Z3016SRV        0  192669552.0                91.0   
148578      Z3016SRV        0  222899312.0                91.0   
183315      Z3016SRV        1   65231360.0                91.0   

        smart_5_normalized  smart_7_normalized  smart_9_normalized  \
9633                 100.0                86.0                72.0   
44370                100.0                86.0                72.0   
79105                100.0                86.0                72.0   
113841               100.0                86.0                72.0   
148578               100.0                86.0                72.0   
183315               100.0                86.0                72.0   

        smart_187_normalized  smart_189_normalized  smart_194_normalized  \
9633                   100.0                  93.0                  25.0   
44370                  100.0                  93.0                  25.0   
79105                  100.0                  93.0                  25.0   
113841                 100.0                  93.0                  24.0   
148578                 100.0                  93.0                  25.0   
183315                 100.0                  93.0                  26.0   

        smart_197_normalized  smart_5_raw  smart_197_raw  
9633                   100.0          0.0           16.0  
44370                  100.0          0.0           24.0  
79105                  100.0          0.0           24.0  
113841                 100.0          0.0           24.0  
148578                 100.0          0.0           32.0  
183315                 100.0          0.0           48.0

In [48]:
def get_ndays_sample(total_df,name_list,new_df,n):
    x=np.array(name_list)
    it = np.nditer(x,flags=['multi_index','refs_ok'])
    groups=total_df.groupby('serial_number')
    while not it.finished:
        ix=it.multi_index
        obj=groups.get_group(x[ix])
        if len(obj)<n:
            it.iternext()
            continue
        obj=obj.iloc[len(obj)-n:len(obj)]
        new_df=new_df.append(obj)
        it.iternext()
    return new_df

In [25]:
def get_healthy_train(total_df,name_list,train_df,n):
    x=np.array(name_list)
    it = np.nditer(x,flags=['multi_index','refs_ok'])
    groups=total_df.groupby('serial_number')
    while not it.finished:
        ix=it.multi_index
        obj=groups.get_group(x[ix])
        train_size=int(len(obj)*0.7)
        if train_size<n:
            it.iternext()
            continue
        obj=obj.iloc[0:train_size+1]
        samples=obj.sample(12)
        train_df=train_df.append(samples)
        it.iternext()
    return train_df

In [15]:
healthy_disk=stdf[~stdf['serial_number'].isin(fail_disk_list['serial_number'])]
len(healthy_disk['serial_number'].unique())

34131

In [17]:
new_disks_list=healthy_disk['serial_number'].unique()
new_disks_list=pd.Series(new_disks_list).sample(frac=0.1)
new_disks_list

20213    S300XPAS
29053    S300XNYV
18354    W300SPHE
17612    Z305D6HR
23138    Z3029FM3
32689    W300BH1H
10042    Z300ZTT1
33890    Z302KJSG
5610     Z300H34V
9597     S301PTEZ
11174    Z302FCX1
13642    W300D52W
13846    Z300ZTSK
2108     Z302F09W
21914    Z300PLDB
17909    S300Z4S4
5709     S301GR4M
21150    Z305D561
32593    Z30251NC
13048    S300Z7B7
18398    Z30427BQ
24473    Z300W1Y0
1112     S300Z7SY
33865    W301D11M
13018    S300ZQP7
28760    Z300C8AL
24629    Z304JL4X
10837    S300Z57X
25016    Z304JXZ8
12078    Z30149WP
           ...   
28127    W3006BS8
20140    Z304HVC2
14915    Z304HWZP
6211     S300Z5WP
20677    Z304LEKX
1323     Z300W2DE
27529    Z300X5LG
19771    Z304JLT3
11144    Z3025209
33916    Z3049FVX
26982    S300WDVS
29269    W300KLMN
806      Z305FJJL
32685    W300TZ62
15212    Z305DVBL
9694     Z3025K9D
11189    Z304HVZJ
16174    Z305B6WZ
33379    W300919Z
1650     Z305FYXP
29665    S301K16Z
27766    Z303VNH7
33212    Z30259ML
11513    Z300PLQZ
16707    Z

In [18]:
new_disks_list.to_csv('/Users/zhilinyang/Desktop/work/new_list.csv')
print('done')

done


In [23]:
healthy_disk.head()

serial_number  failure  smart_1_raw  smart_3_normalized  smart_5_normalized  \
0      Z305B2QN        0   58173272.0                91.0               100.0   
1      Z302A0YH        0   75626904.0                92.0               100.0   
2      Z305BT0W        0   48893128.0                93.0               100.0   
3      Z302A0YE        0  192617016.0                92.0               100.0   
4      Z302PGH8        0   83211536.0                92.0               100.0   

   smart_7_normalized  smart_9_normalized  smart_187_normalized  \
0                85.0                90.0                 100.0   
1                88.0                81.0                 100.0   
2                84.0                92.0                 100.0   
3                75.0                81.0                 100.0   
4                87.0                85.0                 100.0   

   smart_189_normalized  smart_194_normalized  smart_197_normalized  \
0                 100.0                  24.0                 100.0   
1                 100.0                  19.0                 100.0   
2                 100.0                  30.0                 100.0   
3                 100.0                  21.0                 100.0   
4                 100.0                  25.0                 100.0   

   smart_5_raw  smart_197_raw  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0

In [26]:
healthy_train=pd.DataFrame()
healthy_train=get_healthy_train(healthy_disk,healthy_disk['serial_number'].unique(),healthy_train,3)

In [30]:
healthy_train.head(10)

serial_number  failure  smart_1_raw  smart_3_normalized  \
1700676      Z305B2QN        0  213728464.0                91.0   
138942       Z305B2QN        0  162706760.0                91.0   
3695146      Z305B2QN        0  162224696.0                91.0   
6356522      Z302A0YH        0  135465552.0                92.0   
3947211      Z302A0YH        0  188265592.0                92.0   
6287704      Z302A0YH        0  243886384.0                92.0   
3498968      Z305BT0W        0  197946640.0                93.0   
7525927      Z305BT0W        0  159811992.0                93.0   
2912169      Z305BT0W        0   37608824.0                93.0   
1215423      Z302A0YE        0   88175816.0                92.0   

         smart_5_normalized  smart_7_normalized  smart_9_normalized  \
1700676               100.0                86.0                89.0   
138942                100.0                85.0                90.0   
3695146               100.0                87.0                87.0   
6356522               100.0                89.0                76.0   
3947211               100.0                89.0                78.0   
6287704               100.0                89.0                76.0   
3498968               100.0                86.0                89.0   
7525927               100.0                87.0                85.0   
2912169               100.0                85.0                89.0   
1215423               100.0                78.0                80.0   

         smart_187_normalized  smart_189_normalized  smart_194_normalized  \
1700676                 100.0                 100.0                  23.0   
138942                  100.0                 100.0                  24.0   
3695146                 100.0                 100.0                  21.0   
6356522                 100.0                 100.0                  21.0   
3947211                 100.0                 100.0                  25.0   
6287704                 100.0                 100.0                  24.0   
3498968                 100.0                 100.0                  25.0   
7525927                 100.0                 100.0                  25.0   
2912169                 100.0                 100.0                  26.0   
1215423                 100.0                 100.0                  17.0   

         smart_197_normalized  smart_5_raw  smart_197_raw  
1700676                 100.0          0.0            0.0  
138942                  100.0          0.0            0.0  
3695146                 100.0          0.0            0.0  
6356522                 100.0          0.0            0.0  
3947211                 100.0          0.0            0.0  
6287704                 100.0          0.0            0.0  
3498968                 100.0          0.0            0.0  
7525927                 100.0          0.0            0.0  
2912169                 100.0          0.0            0.0  
1215423                 100.0          0.0            0.0

In [31]:
len(healthy_train)

102387

In [33]:
def get_healthy_test(total_df,name_list,train_df):
    x=np.array(name_list)
    it = np.nditer(x,flags=['multi_index','refs_ok'])
    groups=total_df.groupby('serial_number')
    while not it.finished:
        ix=it.multi_index
        obj=groups.get_group(x[ix])
        test_size=int(len(obj)*0.3)
        samples=obj.iloc[-1:-1-test_size:-1]
        samples=samples.sample(3)
        train_df=train_df.append(samples)
        it.iternext()
    return train_df

In [34]:
healthy_test=pd.DataFrame()
healthy_test=get_healthy_test(healthy_disk,healthy_train['serial_number'].unique(),healthy_test)

In [36]:
healthy_test.head(15)

serial_number  failure  smart_1_raw  smart_3_normalized  \
10481138      Z305B2QN        0  137250528.0                91.0   
11561857      Z305B2QN        0   69439432.0                91.0   
10579055      Z305B2QN        0  115027136.0                91.0   
8882791       Z302A0YH        0  131242080.0                92.0   
11419093      Z302A0YH        0  191461904.0                92.0   
11548575      Z302A0YH        0   37423544.0                92.0   
10682427      Z305BT0W        0  226525272.0                93.0   
10756459      Z305BT0W        0  197163920.0                93.0   
10723867      Z305BT0W        0    9921776.0                93.0   
9486004       Z302A0YE        0  232262568.0                92.0   
9748221       Z302A0YE        0   65692120.0                92.0   
10009952      Z302A0YE        0  237501664.0                92.0   
11933463      Z302PGH8        0   48412528.0                92.0   
8947973       Z302PGH8        0  216850784.0                92.0   
11189624      Z302PGH8        0  157866712.0                92.0   

          smart_5_normalized  smart_7_normalized  smart_9_normalized  \
10481138               100.0                88.0                82.0   
11561857               100.0                88.0                81.0   
10579055               100.0                88.0                81.0   
8882791                100.0                89.0                74.0   
11419093               100.0                90.0                72.0   
11548575               100.0                90.0                72.0   
10682427               100.0                88.0                83.0   
10756459               100.0                88.0                83.0   
10723867               100.0                88.0                83.0   
9486004                100.0                83.0                73.0   
9748221                100.0                83.0                73.0   
10009952               100.0                83.0                73.0   
11933463               100.0                89.0                75.0   
8947973                100.0                89.0                77.0   
11189624               100.0                89.0                76.0   

          smart_187_normalized  smart_189_normalized  smart_194_normalized  \
10481138                 100.0                 100.0                  23.0   
11561857                 100.0                 100.0                  23.0   
10579055                 100.0                 100.0                  22.0   
8882791                  100.0                 100.0                  19.0   
11419093                 100.0                 100.0                  18.0   
11548575                 100.0                 100.0                  26.0   
10682427                 100.0                 100.0                  25.0   
10756459                 100.0                 100.0                  26.0   
10723867                 100.0                 100.0                  26.0   
9486004                  100.0                 100.0                  18.0   
9748221                  100.0                 100.0                  19.0   
10009952                 100.0                 100.0                  18.0   
11933463                 100.0                 100.0                  23.0   
8947973                  100.0                 100.0                  21.0   
11189624                 100.0                 100.0                  25.0   

          smart_197_normalized  smart_5_raw  smart_197_raw  
10481138                 100.0          0.0            0.0  
11561857                 100.0          0.0            0.0  
10579055                 100.0          0.0            0.0  
8882791                  100.0          0.0            0.0  
11419093                 100.0          0.0            0.0  
11548575                 100.0          0.0            0.0  
10682427                 100.0          0.0            0.0  
10756459                 100.0

In [ ]:
len(healthy_test)

In [63]:
new_df=pd.DataFrame()
fail_train=get_ndays_sample(stdf,train_disk_list['serial_number'],new_df,12)

In [65]:
fail_train.head(20)

serial_number  failure  smart_1_raw  smart_3_normalized  \
10006044      Z3016WC8        0  134144160.0                93.0   
10038714      Z3016WC8        0  178693336.0                93.0   
10071390      Z3016WC8        0  235259576.0                93.0   
10104102      Z3016WC8        0   22708792.0                93.0   
10136811      Z3016WC8        0   42442792.0                93.0   
10169523      Z3016WC8        0  142492240.0                93.0   
10202229      Z3016WC8        0  181895144.0                93.0   
10234935      Z3016WC8        0    1884016.0                93.0   
10267635      Z3016WC8        0   72909248.0                93.0   
10300331      Z3016WC8        0  126959568.0                93.0   
10333024      Z3016WC8        0  189773632.0                93.0   
10365710      Z3016WC8        1   56820384.0                93.0   
337159        Z300J9C4        0  200921320.0                92.0   
371897        Z300J9C4        0  240891440.0                92.0   
406638        Z300J9C4        0   47160120.0                92.0   
441376        Z300J9C4        0   91080624.0                92.0   
476116        Z300J9C4        0  144780944.0                92.0   
510853        Z300J9C4        0  181955896.0                92.0   
545592        Z300J9C4        0  226110592.0                92.0   
580331        Z300J9C4        0   27021688.0                92.0   

          smart_5_normalized  smart_7_normalized  smart_9_normalized  \
10006044               100.0                89.0                64.0   
10038714               100.0                89.0                64.0   
10071390               100.0                89.0                64.0   
10104102               100.0                89.0                64.0   
10136811               100.0                89.0                64.0   
10169523               100.0                89.0                64.0   
10202229               100.0                89.0                64.0   
10234935               100.0                89.0                64.0   
10267635               100.0                89.0                64.0   
10300331               100.0                89.0                64.0   
10333024               100.0                89.0                64.0   
10365710               100.0                89.0                64.0   
337159                 100.0                89.0                67.0   
371897                 100.0                89.0                67.0   
406638                 100.0                89.0                67.0   
441376                 100.0                89.0                67.0   
476116                 100.0                89.0                67.0   
510853                 100.0                89.0                67.0   
545592                 100.0                89.0                67.0   
580331                 100.0                89.0                67.0   

          smart_187_normalized  smart_189_normalized  smart_194_normalized  \
10006044                 100.0                  96.0                  20.0   
10038714                 100.0                  96.0                  20.0   
10071390                 100.0                  96.0                  20.0   
10104102                 100.0                  96.0                  20.0   
10136811                 100.0                  96.0                  20.0   
10169523                 100.0                  96.0                  19.0   
10202229                 100.0                  96.0                  19.0   
10234935                 100.0                  96.0                  20.0   
10267635                 100.0                  96.0                  19.0   
10300331                 100.0                  96.0                  19.0   
10333024                 100.0                  96.0                  20.0   
10365710                 100.0                  96.0                  20.0   
337159                   100.0                  99.0 

In [77]:
new_df=pd.DataFrame()
fail_test=get_ndays_sample(stdf,test_disk_list['serial_number'],new_df,12)

In [69]:
len(train_disk_list['serial_number'])

743

In [78]:
fail_test.head()

serial_number  failure  smart_1_raw  smart_3_normalized  \
44733       Z3015BZD        0  137666992.0                92.0   
79467       Z3015BZD        0  174417144.0                92.0   
114203      Z3015BZD        0  224017144.0                92.0   
148940      Z3015BZD        0   20801200.0                92.0   
183677      Z3015BZD        0   60922240.0                92.0   

        smart_5_normalized  smart_7_normalized  smart_9_normalized  \
44733                100.0                87.0                73.0   
79467                100.0                87.0                73.0   
114203               100.0                87.0                73.0   
148940               100.0                87.0                73.0   
183677               100.0                87.0                73.0   

        smart_187_normalized  smart_189_normalized  smart_194_normalized  \
44733                  100.0                  95.0                  27.0   
79467                  100.0                  95.0                  27.0   
114203                 100.0                  95.0                  26.0   
148940                 100.0                  95.0                  27.0   
183677                 100.0                  95.0                  27.0   

        smart_197_normalized  smart_5_raw  smart_197_raw  
44733                  100.0          0.0            0.0  
79467                  100.0          0.0            0.0  
114203                 100.0          0.0            0.0  
148940                 100.0          0.0            0.0  
183677                 100.0          0.0            0.0

In [81]:
fail_train.to_csv('/Users/zhilinyang/Desktop/failed_train.csv')
print('done')

done


In [82]:
fail_test.to_csv('/Users/zhilinyang/Desktop/failed_test.csv')
print('done')

done


In [8]:
fail_train=pd.read_csv('/Users/zhilinyang/Desktop/failed_train.csv')
fail_test=pd.read_csv('/Users/zhilinyang/Desktop/failed_test.csv')
fail_train.head()

Unnamed: 0 serial_number  failure  smart_1_raw  smart_3_normalized  \
0    10006044      Z3016WC8        0  134144160.0                93.0   
1    10038714      Z3016WC8        0  178693336.0                93.0   
2    10071390      Z3016WC8        0  235259576.0                93.0   
3    10104102      Z3016WC8        0   22708792.0                93.0   
4    10136811      Z3016WC8        0   42442792.0                93.0   

   smart_5_normalized  smart_7_normalized  smart_9_normalized  \
0               100.0                89.0                64.0   
1               100.0                89.0                64.0   
2               100.0                89.0                64.0   
3               100.0                89.0                64.0   
4               100.0                89.0                64.0   

   smart_187_normalized  smart_189_normalized  smart_194_normalized  \
0                 100.0                  96.0                  20.0   
1                 100.0                  96.0                  20.0   
2                 100.0                  96.0                  20.0   
3                 100.0                  96.0                  20.0   
4                 100.0                  96.0                  20.0   

   smart_197_normalized  smart_5_raw  smart_197_raw  
0                 100.0          0.0           72.0  
1                 100.0          0.0           88.0  
2                 100.0          0.0           88.0  
3                 100.0          0.0           88.0  
4                 100.0          0.0           88.0

In [9]:
fail_train=fail_train.iloc[:,1:]
fail_train.head(24)

serial_number  failure  smart_1_raw  smart_3_normalized  \
0       Z3016WC8        0  134144160.0                93.0   
1       Z3016WC8        0  178693336.0                93.0   
2       Z3016WC8        0  235259576.0                93.0   
3       Z3016WC8        0   22708792.0                93.0   
4       Z3016WC8        0   42442792.0                93.0   
5       Z3016WC8        0  142492240.0                93.0   
6       Z3016WC8        0  181895144.0                93.0   
7       Z3016WC8        0    1884016.0                93.0   
8       Z3016WC8        0   72909248.0                93.0   
9       Z3016WC8        0  126959568.0                93.0   
10      Z3016WC8        0  189773632.0                93.0   
11      Z3016WC8        1   56820384.0                93.0   
12      Z300J9C4        0  200921320.0                92.0   
13      Z300J9C4        0  240891440.0                92.0   
14      Z300J9C4        0   47160120.0                92.0   
15      Z300J9C4        0   91080624.0                92.0   
16      Z300J9C4        0  144780944.0                92.0   
17      Z300J9C4        0  181955896.0                92.0   
18      Z300J9C4        0  226110592.0                92.0   
19      Z300J9C4        0   27021688.0                92.0   
20      Z300J9C4        0   64580704.0                92.0   
21      Z300J9C4        0  101099840.0                92.0   
22      Z300J9C4        0  139038632.0                92.0   
23      Z300J9C4        1  179359256.0                92.0   

    smart_5_normalized  smart_7_normalized  smart_9_normalized  \
0                100.0                89.0                64.0   
1                100.0                89.0                64.0   
2                100.0                89.0                64.0   
3                100.0                89.0                64.0   
4                100.0                89.0                64.0   
5                100.0                89.0                64.0   
6                100.0                89.0                64.0   
7                100.0                89.0                64.0   
8                100.0                89.0                64.0   
9                100.0                89.0                64.0   
10               100.0                89.0                64.0   
11               100.0                89.0                64.0   
12               100.0                89.0                67.0   
13               100.0                89.0                67.0   
14               100.0                89.0                67.0   
15               100.0                89.0                67.0   
16               100.0                89.0                67.0   
17               100.0                89.0                67.0   
18               100.0                89.0                67.0   
19               100.0                89.0                67.0   
20               100.0                89.0                67.0   
21               100.0                89.0                67.0   
22               100.0                89.0                67.0   
23               100.0                89.0                67.0   

    smart_187_normalized  smart_189_normalized  smart_194_normalized  \
0                  100.0                  96.0                  20.0   
1                  100.0                  96.0                  20.0   
2                  100.0                  96.0                  20.0   
3                  100.0                  96.0                  20.0   
4                  100.0                  96.0                  20.0   
5                  100.0                  96.0                  19.0   
6                  100.0                  96.0                  19.0   
7                  100.0                  96.0                  20.0   
8                  100.0                  96.0                  19.0   
9                  100.0                  96.0                  19.0   
10       

In [ ]:
fail_test=fail_test.iloc[:,1:]

In [12]:
fail_train=fail_train.drop('failure',axis=1)
fail_test=fail_test.drop('failure',axis=1)
fail_test.head()

serial_number  smart_1_raw  smart_3_normalized  smart_5_normalized  \
0      Z3015BZD  137666992.0                92.0               100.0   
1      Z3015BZD  174417144.0                92.0               100.0   
2      Z3015BZD  224017144.0                92.0               100.0   
3      Z3015BZD   20801200.0                92.0               100.0   
4      Z3015BZD   60922240.0                92.0               100.0   

   smart_7_normalized  smart_9_normalized  smart_187_normalized  \
0                87.0                73.0                 100.0   
1                87.0                73.0                 100.0   
2                87.0                73.0                 100.0   
3                87.0                73.0                 100.0   
4                87.0                73.0                 100.0   

   smart_189_normalized  smart_194_normalized  smart_197_normalized  \
0                  95.0                  27.0                 100.0   
1                  95.0                  27.0                 100.0   
2                  95.0                  26.0                 100.0   
3                  95.0                  27.0                 100.0   
4                  95.0                  27.0                 100.0   

   smart_5_raw  smart_197_raw  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0

In [37]:
healthy_train=healthy_train.drop('failure',axis=1)
healthy_test=healthy_test.drop('failure',axis=1)
healthy_test.head()

serial_number  smart_1_raw  smart_3_normalized  smart_5_normalized  \
10481138      Z305B2QN  137250528.0                91.0               100.0   
11561857      Z305B2QN   69439432.0                91.0               100.0   
10579055      Z305B2QN  115027136.0                91.0               100.0   
8882791       Z302A0YH  131242080.0                92.0               100.0   
11419093      Z302A0YH  191461904.0                92.0               100.0   

          smart_7_normalized  smart_9_normalized  smart_187_normalized  \
10481138                88.0                82.0                 100.0   
11561857                88.0                81.0                 100.0   
10579055                88.0                81.0                 100.0   
8882791                 89.0                74.0                 100.0   
11419093                90.0                72.0                 100.0   

          smart_189_normalized  smart_194_normalized  smart_197_normalized  \
10481138                 100.0                  23.0                 100.0   
11561857                 100.0                  23.0                 100.0   
10579055                 100.0                  22.0                 100.0   
8882791                  100.0                  19.0                 100.0   
11419093                 100.0                  18.0                 100.0   

          smart_5_raw  smart_197_raw  
10481138          0.0            0.0  
11561857          0.0            0.0  
10579055          0.0            0.0  
8882791           0.0            0.0  
11419093          0.0            0.0

In [39]:
print(len(healthy_train),len(healthy_test),len(fail_train),len(fail_test))

102387 102387 8772 3648


In [48]:
train_x=healthy_train.append(fail_train)
train_x.index=range(0,len(train_x))
train_y=[0 for v in range(0,len(healthy_train))]+[1 for v in range(0,len(fail_train))]
train_y=pd.Series(train_y)

test_x=healthy_test.append(fail_test)
test_x.index=range(0,len(test_x))
test_y=[0 for v in range(0,len(healthy_test))]+[1 for v in range(0,len(fail_test))]
test_y=pd.Series(test_y)

train_df=pd.concat([train_x,train_y],axis=1).sample(frac=1)
test_df=pd.concat([test_x,test_y],axis=1).sample(frac=1)
test_df.head(15)

serial_number  smart_1_raw  smart_3_normalized  smart_5_normalized  \
14927       W300R8FD  210428384.0                96.0               100.0   
28294       Z304JMTY  185917944.0                97.0               100.0   
102896      Z301JZLE  235594800.0                93.0               100.0   
63579       S300XE2H   79078632.0               100.0               100.0   
64897       S300Z5LH  167795152.0                97.0               100.0   
93622       S300Z6SR  121069056.0                95.0               100.0   
13162       Z300XG68   61975456.0                94.0               100.0   
16858       Z301HR3B  227353776.0                92.0               100.0   
12938       S300Z6AZ  123836144.0                94.0               100.0   
70213       Z305K20G   72089464.0                97.0               100.0   
10026       S301LSRN  147140680.0                92.0               100.0   
60507       Z305GXLF   24505928.0                96.0               100.0   
37703       Z305DFVV  123506440.0                96.0               100.0   
1309        W3009ZN1   86468424.0                91.0               100.0   
50727       Z304KEW4  137319656.0                93.0               100.0   

        smart_7_normalized  smart_9_normalized  smart_187_normalized  \
14927                 72.0                75.0                 100.0   
28294                 88.0                82.0                 100.0   
102896                84.0                90.0                 100.0   
63579                 90.0                76.0                 100.0   
64897                 79.0                76.0                 100.0   
93622                 90.0                75.0                 100.0   
13162                 78.0                61.0                 100.0   
16858                 89.0                79.0                 100.0   
12938                 88.0                75.0                 100.0   
70213                 88.0                83.0                 100.0   
10026                 87.0                82.0                 100.0   
60507                 88.0                84.0                 100.0   
37703                 83.0                90.0                 100.0   
1309                  84.0                56.0                 100.0   
50727                 89.0                79.0                 100.0   

        smart_189_normalized  smart_194_normalized  smart_197_normalized  \
14927                  100.0                  21.0                 100.0   
28294                  100.0                  22.0                 100.0   
102896                 100.0                  22.0                 100.0   
63579                  100.0                  20.0                 100.0   
64897                  100.0                  27.0                 100.0   
93622                  100.0                  19.0                 100.0   
13162                   89.0                  25.0                 100.0   
16858                  100.0                  30.0                 100.0   
12938                  100.0                  22.0                 100.0   
70213                  100.0                  32.0                 100.0   
10026                  100.0                  24.0                 100.0   
60507                  100.0                  25.0                 100.0   
37703                  100.0                  34.0                 100.0   
1309                    99.0                  24.0                 100.0   
50727                  100.0                  25.0                 100.0   

        smart_5_raw  smart_197_raw  0  
14927           0.0            0.0  0  
28294           0.0            0.0  0  
102896          0.0            0.0  1  
63579           0.0            0.0  0  
64897           0.0            0.0  0  
93622           0.0            0.0  0  
13162           0.0            0.0  0  
16858           0.0            0.0  0  
12938           0.0            0.0  0  
7021

In [50]:
x_train=train_df.iloc[:,:-1]
x_train.head()

serial_number  smart_1_raw  smart_3_normalized  smart_5_normalized  \
34198      Z304HVAM   74048288.0                95.0               100.0   
22118      S300XRA6  100065256.0                96.0               100.0   
57810      Z304JMSQ   75736472.0                92.0               100.0   
88686      Z302B0HC  242383000.0                92.0               100.0   
9584       Z300MG4B   44841392.0                92.0               100.0   

       smart_7_normalized  smart_9_normalized  smart_187_normalized  \
34198                84.0                89.0                 100.0   
22118                89.0                78.0                 100.0   
57810                87.0                85.0                 100.0   
88686                89.0                78.0                 100.0   
9584                 75.0                64.0                 100.0   

       smart_189_normalized  smart_194_normalized  smart_197_normalized  \
34198                 100.0                  26.0                 100.0   
22118                 100.0                  19.0                 100.0   
57810                 100.0                  27.0                 100.0   
88686                 100.0                  30.0                 100.0   
9584                   88.0                  22.0                 100.0   

       smart_5_raw  smart_197_raw  
34198          0.0            0.0  
22118          0.0            0.0  
57810          0.0            0.0  
88686          0.0            0.0  
9584           0.0            0.0

In [51]:
x_test=test_df.iloc[:,:-1]
y_train=train_df.iloc[:,-1]
y_test=test_df.iloc[:,-1]
y_train.head()

34198    0
22118    0
57810    0
88686    0
9584     0
Name: 0, dtype: int64

In [52]:
x_train.to_csv('/Users/zhilinyang/Desktop/work/x_train1.csv')
print('done')

done


In [53]:
x_test.to_csv('/Users/zhilinyang/Desktop/work/x_test1.csv')
print('done')

done


In [54]:
y_train.to_csv('/Users/zhilinyang/Desktop/work/y_train1.csv')
print('done')

done


In [55]:
y_test.to_csv('/Users/zhilinyang/Desktop/work/y_test1.csv')
print('done')

done


In [ ]:
def feature_extraction(dataset,features):
    return dataset[features]

In [ ]:
def change_rate(features,n,dataset,newdf):
    for feature in features:
        a=dataset[feature][1:]
        b=dataset[feature][:-1]
        c=a-b
        newdf=pd.concat([newdf,c],axis=1)
    dataset=pd.concat([dataset,newdf],axis=1)
    return dataset

In [2]:
stdf=pd.read_csv('/Users/zhilinyang/Desktop/work/STA.csv')
stdf.head()

serial_number  failure  smart_1_raw  smart_3_normalized  smart_5_normalized  \
0      Z305B2QN        0   58173272.0                91.0               100.0   
1      Z302A0YH        0   75626904.0                92.0               100.0   
2      Z305BT0W        0   48893128.0                93.0               100.0   
3      Z302A0YE        0  192617016.0                92.0               100.0   
4      Z302PGH8        0   83211536.0                92.0               100.0   

   smart_7_normalized  smart_9_normalized  smart_187_normalized  \
0                85.0                90.0                 100.0   
1                88.0                81.0                 100.0   
2                84.0                92.0                 100.0   
3                75.0                81.0                 100.0   
4                87.0                85.0                 100.0   

   smart_189_normalized  smart_194_normalized  smart_197_normalized  \
0                 100.0                  24.0                 100.0   
1                 100.0                  19.0                 100.0   
2                 100.0                  30.0                 100.0   
3                 100.0                  21.0                 100.0   
4                 100.0                  25.0                 100.0   

   smart_5_raw  smart_197_raw  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0

In [3]:
len(stdf['serial_number'].unique())

35189

In [4]:
35189-1058

34131

In [8]:
groups=stdf.groupby('serial_number')
groups.get_group('Z305BT0W')

serial_number  failure  smart_1_raw  smart_3_normalized  \
2             Z305BT0W        0   48893128.0                93.0   
34739         Z305BT0W        0  173666208.0                93.0   
69475         Z305BT0W        0  103793736.0                93.0   
104207        Z305BT0W        0  213479224.0                93.0   
138944        Z305BT0W        0  231123768.0                93.0   
173680        Z305BT0W        0  189815464.0                93.0   
208417        Z305BT0W        0  199421232.0                93.0   
243154        Z305BT0W        0  135533416.0                93.0   
277891        Z305BT0W        0  115197056.0                93.0   
312628        Z305BT0W        0   62564872.0                93.0   
347366        Z305BT0W        0  103345136.0                93.0   
382104        Z305BT0W        0   68174720.0                93.0   
416842        Z305BT0W        0  203637648.0                93.0   
451580        Z305BT0W        0  188586832.0                93.0   
486318        Z305BT0W        0   70203456.0                93.0   
521056        Z305BT0W        0   26673296.0                93.0   
555795        Z305BT0W        0   19896504.0                93.0   
590532        Z305BT0W        0   47504352.0                93.0   
625266        Z305BT0W        0  101707400.0                93.0   
660000        Z305BT0W        0  212511904.0                93.0   
694730        Z305BT0W        0  162902912.0                93.0   
729459        Z305BT0W        0  109265888.0                93.0   
764186        Z305BT0W        0   57156208.0                93.0   
798909        Z305BT0W        0  143828712.0                93.0   
833631        Z305BT0W        0   40333064.0                93.0   
868353        Z305BT0W        0   12997368.0                93.0   
903074        Z305BT0W        0   12997368.0                93.0   
937795        Z305BT0W        0  119535272.0                93.0   
972502        Z305BT0W        0   75858696.0                93.0   
1007208       Z305BT0W        0  213077408.0                93.0   
...                ...      ...          ...                 ...   
11275875      Z305BT0W        0  185619968.0                93.0   
11308290      Z305BT0W        0  173709192.0                93.0   
11340704      Z305BT0W        0   12011408.0                93.0   
11373118      Z305BT0W        0   87560416.0                93.0   
11405532      Z305BT0W        0  177300896.0                93.0   
11437946      Z305BT0W        0  143702984.0                93.0   
11470358      Z305BT0W        0  118056704.0                93.0   
11502764      Z305BT0W        0   26814936.0                93.0   
11535118      Z305BT0W        0  161982320.0                93.0   
11567341      Z305BT0W        0   73655512.0                93.0   
11599714      Z305BT0W        0  145673184.0                93.0   
11631975      Z305BT0W        0   53273912.0                93.0   
11664319      Z305BT0W        0  202122640.0                93.0   
11696664      Z305BT0W        0  188164472.0                93.0   
11728954      Z305BT0W        0   94449224.0                93.0   
11761278      Z305BT0W        0   92742192.0                93.0   
11793601      Z305BT0W        0  188556656.0                93.0   
11825924      Z305BT0W        0  121859936.0                93.0   
11858223      Z305BT0W        0   71525960.0                92.0   
11890398      Z305BT0W        0  212371280.0                92.0   
11922536      Z305BT0W        0   45245016.0                92.0   
11954652      Z305BT0W        0  127410344.0                92.0   
11986660      Z305BT0W        0  140682408.0                92.0   
12018744      Z305BT0W        0   69439728.0                92.0   
12050821      Z305BT0W        0  132235552.0                92.0   
12082897      Z305BT0W        0   95649440.0                92.0   
12114971      Z305BT0W        0  214231096.0                92.0 

In [9]:
groups.get_group('Z301JZLE').describe


serial_number  failure  smart_1_raw  smart_3_normalized  \
5089         Z301JZLE        0  207528304.0                93.0   
39826        Z301JZLE        0    1292520.0                93.0   
74562        Z301JZLE        0  103791808.0                93.0   
109298       Z301JZLE        0  111280784.0                93.0   
144035       Z301JZLE        0  243271200.0                93.0   
178771       Z301JZLE        0  124506328.0                93.0   
213508       Z301JZLE        0   36198832.0                93.0   
248244       Z301JZLE        0  225882232.0                93.0   
282981       Z301JZLE        0   45999240.0                93.0   
317719       Z301JZLE        0   44347936.0                93.0   
352457       Z301JZLE        0  154469576.0                93.0   
387198       Z301JZLE        0  235090136.0                93.0   
421936       Z301JZLE        0  137147160.0                93.0   
456674       Z301JZLE        0  114040160.0                93.0   
491411       Z301JZLE        0   99805920.0                93.0   
526148       Z301JZLE        0  157851168.0                93.0   
560886       Z301JZLE        0   24219536.0                93.0   
595623       Z301JZLE        0  225069776.0                93.0   
630357       Z301JZLE        0  146549232.0                93.0   
665091       Z301JZLE        0   51852184.0                93.0   
699820       Z301JZLE        0  152799488.0                93.0   
734549       Z301JZLE        0   46562088.0                93.0   
769275       Z301JZLE        0  214364352.0                93.0   
803998       Z301JZLE        0   73801280.0                93.0   
838720       Z301JZLE        0  208150608.0                93.0   
873441       Z301JZLE        0   48953680.0                93.0   
908162       Z301JZLE        0   48953680.0                93.0   
942877       Z301JZLE        0  180858416.0                93.0   
977584       Z301JZLE        0   16579680.0                93.0   
1012290      Z301JZLE        0   71641024.0                93.0   
...               ...      ...          ...                 ...   
1081699      Z301JZLE        0   46420792.0                93.0   
1116404      Z301JZLE        0  203319344.0                93.0   
1151106      Z301JZLE        0  115563360.0                93.0   
1185807      Z301JZLE        0  118383824.0                93.0   
1220506      Z301JZLE        0  199423288.0                93.0   
1255203      Z301JZLE        0   35635880.0                93.0   
1289852      Z301JZLE        0  151099448.0                93.0   
1324285      Z301JZLE        0   58596016.0                93.0   
1358976      Z301JZLE        0   25150048.0                93.0   
1393655      Z301JZLE        0  233463648.0                93.0   
1428343      Z301JZLE        0  185715672.0                93.0   
1463022      Z301JZLE        0   10520792.0                93.0   
1497673      Z301JZLE        0  113565416.0                93.0   
1532351      Z301JZLE        0  209954040.0                93.0   
1567027      Z301JZLE        0  186686248.0                93.0   
1601703      Z301JZLE        0   50469176.0                93.0   
1636376      Z301JZLE        0  176446872.0                93.0   
1671047      Z301JZLE        0   76562184.0                93.0   
1705716      Z301JZLE        0  206632288.0                93.0   
1740381      Z301JZLE        0   82381408.0                93.0   
1775045      Z301JZLE        0   56460480.0                93.0   
1809707      Z301JZLE        0    5330096.0                93.0   
1844363      Z301JZLE        0  173026920.0                93.0   
1879022      Z301JZLE        0  235594800.0                93.0   
1913675      Z301JZLE        0  205947632.0                93.0   
1948325      Z301JZLE        0    5101520.0                93.0   
1982965      Z301JZLE        0   80109912.0                93.0   
2017606      Z301JZLE        0  203204240.0            

In [6]:
stdf.std(axis=0)

failure                 9.310844e-03
smart_1_raw             7.048288e+07
smart_3_normalized      2.521527e+00
smart_5_normalized      2.023904e-01
smart_7_normalized      4.185268e+00
smart_9_normalized      9.122558e+00
smart_187_normalized    9.315891e-01
smart_189_normalized    6.182678e+00
smart_194_normalized    4.475583e+00
smart_197_normalized    5.350651e-02
smart_5_raw             2.609522e+02
smart_197_raw           9.977192e+00
dtype: float64

In [5]:
print('a')

a
